### DISCUSSION TITLE
Improving Database Performance

### DISCUSSION THREAD

**Discussion by jackbrown1993** [photoprism/photoprism#1768] [Q&A]
Hi all,

I have a large library of over 110,000 photos, and some actions are very slow, such as naming a face which takes 10-20seconds each time.

My originals are stored on a NAS on the same LAN segment, however, photoprism data is on local NVME storage along with the database. The server is running an Intel Atom 3758 8 Core CPU and over 128GB of DDR4 ECC RAM.

Using htop on my server (Ubuntu, Docker install) I can see CPU spikes for process 'photoprism start' and also MySQL when naming faces.

I have had a search through existing issues and see there are some suggestions about increasing innodb size etc but thought it would be worth a discussion first.

Running latest version 211128-7e8974fd-Linux-x86_64

```
time="2021-11-28T23:18:06Z" level=debug msg="covers: updated 0 subjects [808.899551ms]"
time="2021-11-28T23:18:07Z" level=debug msg="counts: updated 1 subject [1.046689947s]"
time="2021-11-28T23:18:07Z" level=debug msg="photo: uid pr3au4117555g0yu title based on 1 person (“<hidden>”)"
time="2021-11-28T23:18:07Z" level=debug msg="photo: uid pr3au4117555g0yu has new title “<hidden> / <hidden> / 2021” [2.278053ms]"
time="2021-11-28T23:18:07Z" level=info msg="photos: found 2 results for id:pr3au4117555g0yu merged:true [2.406506ms]"
time="2021-11-28T23:18:07Z" level=info msg="changes successfully saved"
time="2021-11-28T23:18:07Z" level=debug msg="http: PUT /api/v1/markers/mr3au463iyrjs3g6 (200) [16.07517322s]"
```

**Comment by lastzero** [comment]
How many pictures do you have? What CPU? Does it happen all the time or just in the beginning? Are you using the latest version?

**Reply by jackbrown1993** [reply]
Updated original post, see bottom.

**Reply by lastzero** [reply]
When you have large tables (many files) it may in fact help to increase the InnoDB cache size as described in the MariaDB documentation. Also make sure Docker didn't put the database files on a slow / remote disk for whatever reason.

Also keep in mind:

While the CPU you mention is comparatively fast for a NAS, it's not very fast in absolute terms. The single-thread performance is well behind our 8-year-old home server CPU, which was cheap at the time of purchase. We run two large PhotoPrism instances on it and have absolutely no issues.

![Screenshot 2021-12-01 at 17 47 26](https://user-images.githubusercontent.com/301686/144277290-60d7d52a-ac02-4439-b79a-5b33372f3f64.png)

**Reply by acranox** [reply]
This mostly matches my observations.  I spent a while doing some storage tuning, and mariadb tuning, but never saw much improvement, so I've mostly come to the conclusion that I'm CPU bound.  I also run an Intel Atom (an even older one than what's mentioned here).  That said, one of the reasons I ultimately settled on Photoprism over a couple other options I tried out was that Photoprism had a much quicker interface.

**Reply by jackbrown1993** [reply]
Yep, it does look like a single core is locking up to 100% when I am seeing delays. Although in terms of databases I don't believe the tables are huge at ~100k rows. Queries are quite complex, I suspect some better indexing/tuning could be done on the tables and queries but an easier option is probably throwing more CPU at this issue. Thanks for looking into this.

**Comment by jackbrown1993** [comment]
Some more data just for notes/future reference, here is an example of some slow queries.
![12_59_00](https://user-images.githubusercontent.com/6218551/144764346-66306776-3643-4fd4-9865-00ecec0f51cf.jpg)

**Reply by srett** [reply]
Regarding those queries in the format of `UPDATE photos SET checked_at .... WHERE id IN (SELECT ..... GROUP BY ....)`: I don't quite understand why they are so slow/have so many Rows_examined. `marker_uid` is the primary key of `markers`, so obviously that should just yield at most 1 row and use an index for lookup; then we join on `files` using `file_uid`, which again is the primary key of `files`, so we again can have at most 1 row and use an index. Then this one `photo_id` is passed to the outer `UPDATE`. I guess you could rewrite this as a single `UPDATE` without the nested select, but I still don't see why it's that slow. Unless I'm missing something very obvious here.

**Reply by lastzero** [reply]
Maybe GROUP BY could be removed from the subselect as it's not strictly necessary... there was an issue with the MariaDB query engine with too many IDs in a subselect. UPDATE ... WHERE is necessary to prevent triggering triggers on non affected rows, which may be executed even if no values change. Right now, we focus on multi-user support, so SQL optimizations have to wait until that is done.

**Reply by srett** [reply]
Yes, at least in this specific case, the inner select can never have more than one result, so it can be removed, but I just tried adding `EXPLAIN` to the inner query and I see no difference at all when trying with or without group by. I'ts already saying it's using index lookups. So maybe something breaks regarding optimization in the UPDATE context.

Understood that multi-user is taking top prio right now; I'm actually really excited for it too, it's the last major feature missing to me.

EDIT: Ok tried analyzing the whole query. Indeed, it's looking at all the rows for the `UPDATE`:
```
MariaDB [photoprism]> EXPLAIN UPDATE photos SET checked_at = NULL WHERE id IN (SELECT f.photo_id FROM files f JOIN markers m ON m.file_uid = f.file_uid WHERE m.marker_uid = 'mr05ug52vy7ntp55');
+------+--------------------+--------+-------+------------------------------+--------------------+---------+-------+-------+-------------+
| id   | select_type        | table  | type  | possible_keys                | key                | key_len | ref   | rows  | Extra       |
+------+--------------------+--------+-------+------------------------------+--------------------+---------+-------+-------+-------------+
|    1 | PRIMARY            | photos | index | NULL                         | PRIMARY            | 4       | NULL  | 37412 | Using where |
|    2 | DEPENDENT SUBQUERY | m      | const | PRIMARY,idx_markers_file_uid | PRIMARY            | 44      | const |     1 |             |
|    2 | DEPENDENT SUBQUERY | f      | const | uix_files_file_uid           | uix_files_file_uid | 45      | const |     1 |             |
+------+--------------------+--------+-------+------------------------------+--------------------+---------+-------+-------+-------------+
```

vs. the non-nested way:
```
MariaDB [photoprism]> EXPLAIN UPDATE photos p JOIN files f ON f.photo_id = p.id JOIN markers m ON m.file_uid = f.file_uid SET p.checked_at = NULL WHERE m.marker_uid = 'mr05ug52vy7ntp55';
+------+-------------+-------+-------+---------------------------------------+--------------------+---------+-------+------+-------+
| id   | select_type | table | type  | possible_keys                         | key                | key_len | ref   | rows | Extra |
+------+-------------+-------+-------+---------------------------------------+--------------------+---------+-------+------+-------+
|    1 | SIMPLE      | m     | const | PRIMARY,idx_markers_file_uid          | PRIMARY            | 44      | const |    1 |       |
|    1 | SIMPLE      | f     | const | uix_files_file_uid,idx_files_photo_id | uix_files_file_uid | 45      | const |    1 |       |
|    1 | SIMPLE      | p     | const | PRIMARY                               | PRIMARY            | 4       | const |    1 |       |
+------+-------------+-------+-------+---------------------------------------+--------------------+---------+-------+------+-------+
```

Maybe the trigger issue can be worked around somehow later on, to use the latter version of the query.

**Reply by lastzero** [reply]
Do you have data that suggests this is faster? According to the docs, WHERE is needed here.

**Reply by srett** [reply]
> According to the docs, WHERE is needed here.

What exactly do you mean by this?

I didn't benchmark this exact case very well (although trying to optimize some of the PP queries is on my very long TODO list of "things to fiddle around with on a free weekend"), but generally speaking if the `EXPLAIN` output tells you it's looking at a lot of rows (and `ref` is `NULL`) a query will get notably slow with many rows. With my 37k photos, the difference is already noticeable in this quick and dirty test:
```
MariaDB [photoprism]> UPDATE photos p JOIN files f ON f.photo_id = p.id JOIN markers m ON m.file_uid = f.file_uid SET p.checked_at = NULL WHERE m.marker_uid = 'mr05ug52vy7ntp55';
Query OK, 1 row affected (0.006 sec)
Rows matched: 1  Changed: 1  Warnings: 0

# Reset back to non-null
MariaDB [photoprism]> UPDATE photos p JOIN files f ON f.photo_id = p.id JOIN markers m ON m.file_uid = f.file_uid SET p.checked_at = '2021-01-01 00:00:00' WHERE m.marker_uid = 'mr05ug52vy7ntp55';
Query OK, 1 row affected (0.008 sec)
Rows matched: 1  Changed: 1  Warnings: 0

MariaDB [photoprism]> UPDATE photos SET checked_at = NULL WHERE id IN (SELECT f.photo_id FROM files f JOIN markers m ON m.file_uid = f.file_uid WHERE m.marker_uid = 'mr05ug52vy7ntp55' GROUP BY f.photo_id);
Query OK, 1 row affected (0.139 sec)
Rows matched: 1  Changed: 1  Warnings: 0
```
I tried this a couple times and the numbers are in the same ballpark each time.

**Reply by lastzero** [reply]
It means we don't want to trigger the triggers and we don't want to lock the whole table. I currently don't have time to do further research on this. If you find a way to work around the trigger issue, let us know.

**Reply by srett** [reply]
Ok, sounded like you had a specific page/article referring to this very issue in mind.

Can you give me any quick pointers about the trigger issue? At least in my setup, there are no triggers in the `photoprism` database, according to `SHOW TRIGGERS`. I'm not proficient in Go, so I can't seem to figure out where they would be created, if any. I'm definitely willing to put more time into this.

That being said, triggers should ever only run for rows that have been *matched* by an `UPDATE`, so at least in this specific example, everything would be fine as you can tell from the status output. `Rows matched: 1  Changed: 1  Warnings: 0` for both query variants. It might very well cause issues with other queries in PP, constructed similarly, but with more involved conditions, I didn't check these yet. But for this specific case, for all intents and purposes, I'm pretty confident they are identical, unless you were actually working around a bug in some older MariaDB version? But maybe we can wait for a second opinion on this. I've used MariaDB and MySQL for a fairly long time and optimized queries for setups with tables with tens of millions of rows, but this right here was some quick testing/debugging on the side, so I won't stake my life on it. ;)

**Reply by lastzero** [reply]
Yes, as long as there is a WHERE like `WHERE m.marker_uid = 'mr05ug52vy7ntp55'` it should not interfere with triggers. While we currently don't have any, there may be some in the future or a user may have added triggers locally. Here is the code:
- https://github.com/photoprism/photoprism/blob/e14209526586fe5115cc7a3e49aa76481b0a12bc/internal/entity/face.go#L277
- https://github.com/photoprism/photoprism/blob/767e3ca492a609c2121060db3032df326d2c6bbd/internal/entity/subject.go#L393

Note it currently doesn't run different queries for MySQL and SQLite. That means you need to make the query also run on SQLite or use the old one there.

I guess if we proceed in big steps like this, we can open a new issue and implement it straight way...

**Reply by lastzero** [reply]
You're welcome to open a Pull Request for the queries after you tested them on MySQL, MariaDB, and SQLite.

**Reply by srett** [reply]
I've got a dev environment up and running, so the testing can begin.

Did some initial research and found that SQLite does not support JOINs in UPDATEs, but has support for [UPDATE FROM](https://www.sqlite.org/lang_update.html#upfrom) since last year. Not sure however if it's worth rewriting for that too, or just keeping the old queries for SQLite. The performance of the nested SELECT version doesn't seem that slow with SQLite (38ms with my 37k files vs. 130ms with MariaDB), so I didn't bother trying UPDATE FROM yet.

Already found another query:
```
# Query_time: 4.779468  Lock_time: 0.000064  Rows_sent: 0  Rows_examined: 4305684
# Rows_affected: 0  Bytes_sent: 11
SET timestamp=1638901106;
DELETE FROM albums WHERE id 
		IN (SELECT a.id FROM albums a JOIN albums b ON a.album_type = b.album_type 
		AND a.album_type <> 'album' AND a.id > b.id WHERE (a.album_slug = b.album_slug 
		OR a.album_filter = b.album_filter) GROUP BY a.album_uid);
```

That's 4 million rows examined :O – my albums table has 2934 rows.

Usually, these kinds of query can be optimized by getting rid of the OR, i.e. turning it into two queries (either two DELETEs or two nested SELECTs using UNION).
Interestingly, that completely solves the performance issues for SQLite, but for MariaDB, `SELECT a.id FROM albums a JOIN albums b ON a.album_type = b.album_type 
		AND a.album_type <> 'album' AND a.id > b.id WHERE (a.album_filter = b.album_filter) GROUP BY a.album_uid` is still almost as slow, as it's still not using any index.

I'll keep working on this, hopefully arriving at a PR in a few days/weeks, or at least dumping my progress in an issue so it doesn't get lost.

**Reply by srett** [reply]
Added PR #1804. For now only addressed the three UPDATE IN SELECT queries that popped up the most. Everything else seems more involved and might require changes to the DB scheme or at least other tricks, not sure how you feel about this at the current stage.

Regarding the DELETE query from my last comment, the performance issues get much better as soon as you add an index on `album_filter` to the `albums` table. For me it drops from 4.7s to 200ms without changing the query.

Then I can get to 70ms with this change:
```
DELETE FROM albums WHERE id
	IN (SELECT a.id FROM albums a JOIN albums b ON a.album_type = b.album_type AND a.album_type <> 'album' AND a.id > b.id
	WHERE a.album_filter = b.album_filter UNION SELECT a.id FROM albums a JOIN albums b ON a.album_type = b.album_type
	AND a.album_type <> 'album' AND a.id > b.id WHERE a.album_slug = b.album_slug);
```

50ms:
```
DELETE a FROM albums a JOIN
	(SELECT a.id FROM albums a JOIN albums b ON a.album_type = b.album_type AND a.album_type <> 'album' AND a.id > b.id
	WHERE a.album_filter = b.album_filter
	UNION SELECT a.id FROM albums a JOIN albums b ON a.album_type = b.album_type AND a.album_type <> 'album' AND a.id > b.id
	WHERE a.album_slug = b.album_slug) b ON (a.id = b.id);
```

But it's a matter of how far you want to bend over backwards I guess. While really slow, this query didn't seem to run that frequently, just during background maintenance, so it should generally not impact UX.

**Reply by lastzero** [reply]
Getting back to you as soon as possible. Seems best to improve in small steps and not to change too much at the same time. Easier to test & debug!

**Reply by lastzero** [reply]
Created a follow-up issue for this (feedback welcome):
- https://github.com/photoprism/photoprism/issues/1953

**Comment by lastzero** [comment]
Added a section on MariaDB Performance Tuning to our docs:
- https://docs.photoprism.org/getting-started/troubleshooting/#performance-tuning

**Comment by jpbuttet** [comment]
> Hi all,
> 
> I have a large library of over 110,000 photos, and some actions are very slow, such as naming a face which takes 10-20seconds each time.
> 
> My originals are stored on a NAS on the same LAN segment, however, photoprism data is on local NVME storage along with the database. The server is running an Intel Atom 3758 8 Core CPU and over 128GB of DDR4 ECC RAM.
> 
> Using htop on my server (Ubuntu, Docker install) I can see CPU spikes for process 'photoprism start' and also MySQL when naming faces.
> 
> I have had a search through existing issues and see there are some suggestions about increasing innodb size etc but thought it would be worth a discussion first.
> 
> Running latest version 211128-7e8974fd-Linux-x86_64
> 
> ```
> time="2021-11-28T23:18:06Z" level=debug msg="covers: updated 0 subjects [808.899551ms]"
> time="2021-11-28T23:18:07Z" level=debug msg="counts: updated 1 subject [1.046689947s]"
> time="2021-11-28T23:18:07Z" level=debug msg="photo: uid pr3au4117555g0yu title based on 1 person (“<hidden>”)"
> time="2021-11-28T23:18:07Z" level=debug msg="photo: uid pr3au4117555g0yu has new title “<hidden> / <hidden> / 2021” [2.278053ms]"
> time="2021-11-28T23:18:07Z" level=info msg="photos: found 2 results for id:pr3au4117555g0yu merged:true [2.406506ms]"
> time="2021-11-28T23:18:07Z" level=info msg="changes successfully saved"
> time="2021-11-28T23:18:07Z" level=debug msg="http: PUT /api/v1/markers/mr3au463iyrjs3g6 (200) [16.07517322s]"
> ```

I am not sure but looks like adressing a face (naming a face) from this page: (new faces) https://xx.xx.xx.xx./people/new is always fast (2-3 seconds) ....

but addressing a face from this page: https://xx.xx.xx.xx./all?q=face%3Anew (page called after pressing button show all news face)

takes from few seconds up to 20 seconds
maybe it helps

best regards

**Reply by Maltavius** [reply]
I'm at above 30 seconds to assign a new name to a face.
Removing a name, the blue bar at the top goes through in a second or two. Adding a new name it takes the blue bar 15-20SS seconds to reach 95% full, and then up to 30-40 seconds more to reach 100% and allow me to actually do something else in the GUI... The performance is horrific.
![image](https://user-images.githubusercontent.com/10133945/150667319-d8f37ff0-1683-4da0-b928-d61f395558db.png)

I've set the MariaDB Cache to 1GB I'm on a pretty fast Machine with 20GB of memory to spare. 

OS: Manjaro 21.2.1 Qonos
Kernel: x86_64 Linux 5.15.12-1-MANJARO
Disk: 3.5T / 12T (31%)
CPU: Intel Core i7-6700K @ 8x 4.4GHz [37.0°C]
GPU: AMD Radeon RX 5700 XT (NAVI10, DRM 3.42.0, 5.15.12-1-MANJARO, LLVM 13.0.0)
RAM: 11408MiB / 32042MiB

**Reply by lastzero** [reply]
How many pictures? How many faces? What version?

**Reply by Maltavius** [reply]
It's the version that docker compose downloads from 2 days ago .
I have 42000 files in Search and 72000 in originals. Index is complete. Standard 2 worker threads or 4 doesn't make a difference.

It happens in all of the pictures. If there are 1 face or 3 in the image doesn't matter.
I have 11 people registered at the moment but I'm hesitant to add more since performance is so bad.

I'm on the Brave browser if that matters.

**Reply by lastzero** [reply]
Wondering if this is related to https://github.com/photoprism/photoprism/issues/1957

**Reply by acranox** [reply]
Anecdotally, I updated to the latest release a day ago, and also ran a manual `faces optimize` and since then, adding a name to a face seems to be much quicker. It maybe went from 4-8 seconds, down to 1-3.  I don't know which change it was, but there definitely seems to have been an improvement.  Sometimes it still take longer than other times, but overall it feels faster.

**Reply by compbl** [reply]
I am new to PhotoPrism.  I am having the slow blue line and would like to "increase the InnoDB cache size as described in the MariaDB documentation".

Being new to docker how would I do that exactly?

Thanks!!

**Reply by graciousgrey** [reply]
It's described here: https://docs.photoprism.app/getting-started/troubleshooting/performance/ :)

**Reply by afrojewelz** [reply]
i am wondering if make mariadb distribute on multiple independt machines within single threaded instance with 10GBe and nvme disk,would that be able to improve current SQL indexing perfomance? does photoprism will support MariaDB Cluster connect to MariaDB in some magic trick?  or just introduce an redis cluster to cache those critical pressured index in DB?

**Comment by spf13** [comment]
I've had similar issues. I switched from Sqlite to MariaDB thinking it would improve things, and it has by avoided the database locking, but I'm still having many queries over 30 seconds when tagging faces.

Here is one such example:
time="2022-07-11T12:42:28Z" level=debug msg="http: PUT /api/v1/markers/mrdlhje3v20n8tpk (200) [38.20958585s]"

I have a very powerful system running Proxmox with Photoprism and MariaDB each in their own containers. 
My specs:
* 400k+ files
* Dual [Intel Xeon E5-2650L v4](https://www.cpubenchmark.net/cpu.php?cpu=Intel+Xeon+E5-2650L+v4+%40+1.70GHz&id=3054) @ 1.70GHz (24 total cores)
* 192 GB ram

I've followed all of the tips in performance tips and far exceeded the amount of RAM and CPU dedicated to these two containers, but neither seems like the bottleneck. RAM usage on both containers is less than 30% utilized and the CPU doesn't really ever get taxed (unless I'm doing a RAW conversion or similar). 

I'd love to help improve the performance and even contribute to the project but I'm not quite clear where to start beyond identifying the API that is often slow, but sometimes reasonably quick (1 - 2 seconds).

**Reply by lastzero** [reply]
For a single query, that seems pretty long indeed. That particular action might do a little more though like updating related pictures. Requires taking a closer look. 

Just to be sure: How did you migrate from Sqlite? Would it be possible that data types are "wrong" or indexes are missing? MariaDB, for example, can't process rows with `text` columns in memory and always uses temporary tables on disk if there are any. This has been refactored earlier this year, but only if you've been running the migrations for MariaDB specifically. 

Any help with improving performance is much appreciated! 💎

**Reply by spf13** [reply]
I followed the exact steps outlined at https://docs.photoprism.app/getting-started/advanced/databases/
(with one exception, the usernames aren't in sync within the doc. The top has you create user "photoprism" the latter migration command uses "root"). 

I don't see any migration for MariaDB step in this other than porting the exact same structure as Sqlite. 

I suppose it's entirely possible that this approach is missing indexes or has different missing optimizations. I'm relatively new to this software and code base. I have an interest in getting it to work though and loads of experience with Go and databases to contribute.

**Reply by spf13** [reply]
Doing some diligence using the slow query log, there's a ton of queries that are happening that aren't using any indexes. My guess is we are missing a lot of indexes using the approach I used. 


Here are a few select entries from that log:

```

# Time: 220711 15:28:21
# User@Host: photoprism[photoprism] @  [192.168.1.215]
# Thread_id: 3  Schema: photoprism  QC_hit: No
# Query_time: 6.682291  Lock_time: 0.000029  Rows_sent: 76645  Rows_examined: 76646
# Rows_affected: 0  Bytes_sent: 837517312
SET timestamp=1657553301;
SELECT * FROM `markers`  WHERE (marker_invalid = 0 AND marker_type = 'face' AND face_id IN (''));

# Time: 220711 15:28:52
# User@Host: photoprism[photoprism] @  [192.168.1.215]
# Thread_id: 3  Schema: photoprism  QC_hit: No
# Query_time: 0.422753  Lock_time: 0.000101  Rows_sent: 0  Rows_examined: 54532
# Rows_affected: 0  Bytes_sent: 0
SET timestamp=1657553332;
UPDATE subjects LEFT JOIN (
        SELECT m.subj_uid, m.q, MAX(m.thumb) AS marker_thumb FROM markers m
                        WHERE m.subj_uid <> '' AND m.subj_uid IS NOT NULL
                          AND m.marker_invalid = 0 AND m.thumb IS NOT NULL AND m.thumb <> ''
                        GROUP BY m.subj_uid, m.q
                        ) b ON b.subj_uid = subjects.subj_uid
                SET thumb = marker_thumb WHERE subjects.subj_type = 'person' AND thumb_src = '';

# Time: 220711 15:27:11
# User@Host: photoprism[photoprism] @  [192.168.1.215]
# Thread_id: 3  Schema: photoprism  QC_hit: No
# Query_time: 1.009595  Lock_time: 0.000103  Rows_sent: 0  Rows_examined: 107994
# Rows_affected: 1  Bytes_sent: 0
SET timestamp=1657553231;
UPDATE subjects LEFT JOIN (
                SELECT m.subj_uid, COUNT(DISTINCT f.id) AS subj_files, COUNT(DISTINCT f.photo_id) AS subj_pho
tos FROM files f
                        JOIN markers m ON f.file_uid = m.file_uid AND m.subj_uid IS NOT NULL AND m.subj_uid <
> '' AND m.subj_uid IS NOT NULL
                        WHERE m.marker_invalid = 0 AND f.deleted_at IS NULL GROUP BY m.subj_uid
                ) b ON b.subj_uid = subjects.subj_uid
                SET subjects.file_count = CASE WHEN b.subj_files IS NULL THEN 0 ELSE b.subj_files END, 
                        subjects.photo_count = CASE WHEN b.subj_photos IS NULL THEN 0 ELSE b.subj_photos END
                WHERE subj_type = 'person';
```


It also appears there are some oddities in the query construction. I would hope MariaDB would optimize things like `AND face_id IN ('')` but I can't imagine it's ideal to have that in there.

**Reply by lastzero** [reply]
These docs were contributed, so I never tested it. We should add a disclaimer as it seems to be missing. My guess is that the instructions work to some extent and give an acceptable result if you don't have a large index. However, 400k files is a lot.

Before you look any further, I would start here and make sure the MariaDB schema matches that of a fresh, non-migrated installation. Apart from that, there should be many ways to optimize performance! Being able to create static galleries for sharing would be awesome too (big fan of Hugo!).

**Reply by spf13** [reply]
I've spent some time looking around the code base. I found [a mariadb init file](https://github.com/photoprism/photoprism/blob/develop/scripts/sql/mariadb-init.sql) but it doesn't look quite right. Definitely doesn't look like my current one. 

I'm not sure where to find the structure of the SQL if I had just started with a MariaDB instance instead of a sqlite one. 

I also looked over the SQL structure of the migrated one and there are a lot of indexes present but for some reason they aren't being used very much. 

I suppose I could install an empty instance with MariaDB and then port the data over to the new structure... but if I can just do the database initialization vs installing everything that would save some time. 

Any tips where to look?

**Reply by lastzero** [reply]
Adding more database documentation was on my to-do list anyway (and had been there for far too long): 

- https://github.com/photoprism/photoprism/tree/develop/internal/entity/schema
- https://docs.photoprism.app/developer-guide/database/schema/

First, try re-running previous migrations in a terminal before manually comparing your schema as a measure of last resort. To get a list of all migrations (what's in the "Dialect" column for you?):

```
$ photoprism migrations ls

|-----------------|---------|---------------------|---------------------|--------|
|       ID        | Dialect |     Started At      |     Finished At     | Status |
|-----------------|---------|---------------------|---------------------|--------|
| 20211121-094727 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
| 20211124-120008 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
| 20220329-030000 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
| 20220329-040000 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
| 20220329-050000 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
| 20220329-060000 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
| 20220329-061000 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
| 20220329-070000 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
| 20220329-071000 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
| 20220329-080000 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
| 20220329-081000 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
| 20220329-083000 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
| 20220329-090000 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
| 20220329-091000 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
| 20220329-093000 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
| 20220421-200000 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
| 20220521-000001 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
| 20220521-000002 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
| 20220521-000003 | mysql   | 2022-07-12 08:07:35 | 2022-07-12 08:07:35 | OK     |
|-----------------|---------|---------------------|---------------------|--------|
```

To re-run migrations, you simply pass them as arguments to the "migrations run" command:

```
$ photoprism migrations run 20220521-000003

INFO[2022-07-12T11:45:29Z] migrate: 20220521-000003 successful [12.967654ms] 
INFO[2022-07-12T11:45:29Z] migration completed in 40.89123ms            
INFO[2022-07-12T11:45:29Z] closed database connection 
```

**Comment by afrojewelz** [comment]
![屏幕截图 2023-11-20 175752](https://github.com/photoprism/photoprism/assets/1831795/9c4f14b5-4748-45d9-8824-309456e3c1b0)
even i m not index my whole img folder in once,manually assign person to unknown subject is stuckin in transaction too long.
my solution of this is running mariadb/gluster data volume direct in emptyDir/tmpfs.which is run innodb ENGIN in Memorydrive. that is direct MMAP which as long as you not loosing power, your memory of db is intact and latency of writing is almost as fast as your single band memory. the whole experience is much better under current version gorm. since gormv2 will not get adopted shortly(but very lookin fwd promisin).
i m guessin there are much work to optimize the trasaction word in manipulate our metadata of imgs. but i don't write golang ,so my hands are tighted.
after my venture db running in memory, i took some time to write online backup db in memory to disk task to crontab,therefore i can manually control the db of photoprism backup. 
btw:  do not run each task of tag subject to person in transaction causing too much non-sequence io write to the disk the UI has to wait till it's done. that's absurd. ya'al gettin too much worried for loosin some little manual input instead of bunch of large input.is there a solution for UI cache the changes user input when user hit apply to all selected object then commit changes to db,that saves a lot I/O intensive task for db?
